components/ai_chat/resources/page/components/markdown_renderer/index.tsx

```diff
@@ -0,0 +1,108 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import * as React from 'react'
+import Markdown from 'react-markdown'
+
+import styles from './style.module.scss'
+import CaretSVG from '../svg/caret'
+
+const CodeBlock = React.lazy(async () => ({
+  default: (await import('../code_block')).default.Block
+}))
+const CodeInline = React.lazy(async () => ({
+  default: (await import('../code_block')).default.Inline
+}))
+
+const allowedElements = [
+  // Headings
+  'h1',
+  'h2',
+  'h3',
+  'h4',
+
+  // Text formatting
+  'blockquote',
+  'code',
+  'del',
+  'em',
+  'strong',
+  'sup',
+
+  // Lists
+  'li',
+  'ol',
+  'ul',
+
+  // Structural elements
+  'p',
+  'pre',
+  'section',
+
+  // Line elements
+  'br',
+  'hr'
+]
+
+interface MarkdownRendererProps {
+  text: string
+  shouldShowTextCursor: boolean
+}
+
+export default function MarkdownRenderer(mainProps: MarkdownRendererProps) {
+  const [lastLine, setLastLine] = React.useState(1)
+
+  const plugin = React.useCallback(() => {
+    const transformer = (ast: any) => {
+      setLastLine(ast.position.end.line)
+    }
+
+    return transformer
+  }, [])
+
```

Comment:
A few questions about this:

1. Will this callback (and therefore setState) be called during render, and not sometime asynchronously afterwards - that could be a performance / circular issue.
2. Is setState getting called a lot - can it negatively affect performance?
3. We're only showing the carat for paragraphs - we don't want to show it for other tags, e.g. headings?

Reply:
1. i had initial concerns of calling `setState` during render too, but react-markdown doesn’t offer an alternate API to do this efficiently and there haven't been issues so far, my strategy is to address any performance issues as they arise. we’re also gating as much as possible; for example, we avoid reading from the AST if the renderer isn’t allowed to show the cursor. ideally, react-markdown should provide each element’s node position info along with 'totalLines' which can be read in the `components`, which would eliminate the need to maintain state in the parent here.
2. `ast.position.end.line` does not change often as chunks are streaming, so i dont think we have to manually handle if prevState was the same as the new one — 'dont call setState'. react should take care of this. additionally, setState is auto batched in react 18, no matter where they originate from.
3. for inline tags like strong, spans, etc., we won't need to show the typing cursor as they're wrapped in `p`. also, I have noticed li's text also includes `p`, but I'm not sure if the cursor is shown here. for headings, we might need to adjust the cursor's height to accommodate the font size. this is the most repeated and daunting part where each tag will need to be custom handled to show a cursor.

Reply:
> …I have noticed li's text also includes `p`…

Are we already taking this into consideration with styling to cut down on default margins for paragraph elements? Presumably we want list items to be a bit more snug together, vertically.

Reply:
ok, so I have modified the AST to include when and where to show the typing cursor, so we won't be using React state for this.

Reply:
regarding the `li`, I've noticed that if the syntax has two new lines "\n\n" between list items, only then does it render the list's content in a paragraph.

i have aligned the spacing where I can, but I'd really like to merge this and test it out in the wild. we can iron out the details in follow-ups.